File: java/src/io/zihengcat/leetcode/window/MaxConsecutiveOnes.java
package io.zihengcat.leetcode.window;

/**
 * LeetCode 485. Max Consecutive Ones
 * https://leetcode.com/problems/max-consecutive-ones/
 */
public class MaxConsecutiveOnes {
    public static void main(String[] args) {
        // ...
    }
    public int findMaxConsecutiveOnes(int[] nums) {
        if (nums == null || nums.length == 0) {
            return 0;
        }
        int maxLen = 0;
        int wLeft = 0;
        int wRight = 0;
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] == 1) {
                wLeft = i;
                wRight = extendRight(nums, i);
                if (wRight - wLeft + 1 > maxLen) {
                    maxLen = wRight - wLeft + 1;
                }
                i = wRight;
            }
        }
        return maxLen;
    }
    private int extendRight(int[] nums, int i) {
        while (i < nums.length && nums[i] == 1) {
            i++;
        }
        return i - 1;
    }
}

/* EOF */

File: cc/array/MaxConsecutiveOnes.cc
#include <vector>

using namespace std;

/**
 * LeetCode 485. Max Consecutive Ones
 * https://leetcode.com/problems/max-consecutive-ones/
 */
class MaxConsecutiveOnes {
public:
    int findMaxConsecutiveOnes(vector<int>& nums) {
        int mOnes = 0;
        int cOnes = 0;
        int n = nums.size();
        for (int i = 0; i < n; /* ... */) {
            switch (nums[i]) {
            case 1:
                cOnes = 0;
                // Count ones
                while (i < n && nums[i] == 1) {
                    i++;
                    cOnes++;
                }
                mOnes = max(mOnes, cOnes);
                break;
            default:
                i++;
                break;
            }
        }
        return mOnes;
    }
};

int main(int argc, char const *argv[]) {
    // ...
    return 0;
}

/* EOF */
